UART_driver/Core/Inc/CircularBuffer.hpp

```diff
@@ -0,0 +1,34 @@
+#ifndef CIRCULAR_BUFFER_HPP
+#define CIRCULAR_BUFFER_HPP
+
+#include <stdint.h>
+
+class CircularBuffer {
+    public:
+        // Constructor. `buf` is the buffer to be used as a circular buffer
+        CircularBuffer(uint8_t* buf, uint16_t size);
+
+        // Peek at a byte in the buffer
+        uint8_t peek(uint8_t& res, uint16_t dist);
+
+        // Read bytes from the buffer
+        uint8_t read(uint8_t* buf, uint16_t dist);
+
+        // Write byte to the buffer
+        uint8_t write(uint8_t byte);
+
+        // Returns true if the buffer has space
+        bool hasSpace();
```

Comment:
This name sounds kinda werid, maybe consider isFull()? Just a personal thing, if you think this works better than leave it